compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -160,21 +173,18 @@ where
         let key = self.key;
         let state = self.state;
 
-        // Forget ourself so our destructor won't poison the query
-        mem::forget(self);
-
         // Mark as complete before we remove the job from the active state
         // so no other thread can re-execute this query.
         cache.complete(key, result, dep_node_index);
 
         let job = {
             let mut lock = state.active.lock_shard_by_value(&key);
-            match lock.remove(&key).unwrap() {
-                QueryResult::Started(job) => job,
-                QueryResult::Poisoned => panic!(),
-            }
+            lock.remove(&key).unwrap().expect_job(self.query_name)
         };
 
+        // Forget ourself so our destructor won't poison the query
+        mem::forget(self);
```

Comment:
Why did you move this down here? If `expect_job` panics, then it will unwind and execute the destructor, which will cause a second panic.

Reply:
Oh yeah I meant to move that back up. Thanks.